### DISCUSSION TITLE
Evaluate using Profile-Guided Optimization (PGO) for D2 CLI

### DISCUSSION THREAD

**Discussion by zamazan4ik** [terrastruct/d2#2234] [Ideas]
Hi!

Recently I checked Profile-Guided Optimization (PGO) improvements on multiple projects including many compilers - all the results are available in the [awesome-pgo](https://github.com/zamazan4ik/awesome-pgo/) repo. Since PGO has helped in many cases, I think it would be a good idea to try optimizing D2 tools by applying PGO.

I can suggest the following things to do:

* We need to perform PGO benchmarks on D2 (the tool and the library). If it shows improvements - add a note about possible improvements to the documentation. Maybe some work on performance benchmarks would useful before that (it was requested in https://github.com/terrastruct/d2/issues/842)
* It will be a good idea to integrate building with PGO into your CI pipelines. In this case, users will get already PGO-optimized binaries from the website

[Here](https://github.com/zamazan4ik/awesome-pgo/) you can find different materials about PGO: benchmarks for multiple apps (mostly for non-Go applications but it's only for now - may be changed later), examples of how PGO is already integrated with various projects, and some PGO-related pieces of advice.

Regarding Go, I can recommend starting to read about PGO from [here](https://github.com/zamazan4ik/awesome-pgo/blob/main/article/article.md#go) and then continue reading official Go [docs](https://go.dev/doc/pgo) about PGO.

Thank you.

**Comment by alixander** [comment]
Thanks for your efforts in advocating this. You're right that https://github.com/terrastruct/d2/issues/842 should be done prior, in order to know if this method yields gains.
